core/server/common/src/main/java/alluxio/underfs/UfsManager.java

```
@@ -0,0 +1,164 @@
+        return true;
+      }
+
+      if (!(object instanceof Key)) {
+        return false;
+      }
+
+      Key that = (Key) object;
+      return Objects.equal(mAuthority, that.mAuthority) && Objects
+          .equal(mProperties, that.mProperties) && Objects.equal(mScheme, that.mScheme);
+    }
+
+    @Override
+    public String toString() {
+      return Objects.toStringHelper(this)
+          .add("authority", mAuthority)
+          .add("scheme", mScheme)
+          .add("properties", mProperties).toString();
+    }
+  }
+
+  /** Maps from key to {@link UnderFileSystem} instances. */
+  private final ConcurrentHashMap<Key, UnderFileSystem> mUnderFileSystemMap =
+      new ConcurrentHashMap<>();
+  /** Maps from mount id to {@link UnderFileSystem} instances. */
+  private final ConcurrentHashMap<Long, UnderFileSystem> mMountIdToUnderFileSystemMap =
+      new ConcurrentHashMap<>();
+
+  private UnderFileSystem mRootUfs = null;
+  protected final Closer mCloser;
+
+  /**
+   * Constructs the instance.
+   */
+  public UfsManager() {
+    mCloser = Closer.create();
+  }
+
+  /**
+   * Gets a UFS instance from the cache if exists. Otherwise, creates a new instance and adds
+   * that to the cache. Use this method only when you create new UFS instances.
+   *
+   * @param path the UFS path
+   * @param ufsConf the UFS configuration
+   * @return the UFS instance
+   */
+  public UnderFileSystem getOrCreate(String path, Map<String, String> ufsConf) {
+    Key key = new Key(new AlluxioURI(path), ufsConf);
+    UnderFileSystem cachedFs = mUnderFileSystemMap.get(key);
+    if (cachedFs != null) {
+      return cachedFs;
+    }
+    UnderFileSystem fs = UnderFileSystemRegistry.create(path, ufsConf);
+    UnderFileSystem racingFs = mUnderFileSystemMap.putIfAbsent(key, fs);
+    if (racingFs == null) {
+      mCloser.register(fs);
+      return fs;
+    }
+    try {
+      fs.close();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```

yeah, i think i wrote this. I think it is better to swallow it and log the exception too